sdk/messaging/azwebpubsub/client_custom.go

```diff
@@ -0,0 +1,185 @@
+// Licensed under the MIT License.
+
+package azwebpubsub
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azwebpubsub/internal"
+	"github.com/golang-jwt/jwt"
+)
+
+// ClientOptions contains optional settings for [Client]
+type ClientOptions struct {
+	azcore.ClientOptions
+}
+
+// NewClient creates a client that manages Web PubSub service
+func NewClient(endpoint string, hub string, credential azcore.TokenCredential, options *ClientOptions) (*Client, error) {
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	authPolicy := runtime.NewBearerTokenPolicy(credential, []string{internal.TokenScope}, nil)
+	azcoreClient, err := azcore.NewClient(internal.ModuleName+".Client", internal.ModuleVersion,
+		runtime.PipelineOptions{PerRetry: []policy.Policy{authPolicy}}, &options.ClientOptions)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		internal: azcoreClient,
+		endpoint: endpoint,
+		hub:      hub,
+	}, nil
+}
+
+// NewClientFromConnectionString creates a Client from a connection string
+func NewClientFromConnectionString(connectionString string, hub string, options *ClientOptions) (*Client, error) {
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	props, err := internal.ParseConnectionString(connectionString)
+
+	if err != nil {
+		return nil, err
+	}
+
+	azcoreClient, err := azcore.NewClient(internal.ModuleName+".Client", internal.ModuleVersion, runtime.PipelineOptions{
+		PerRetry: []policy.Policy{internal.NewWebPubSubKeyCredentialPolicy(props.AccessKey)},
+	}, &options.ClientOptions)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		internal: azcoreClient,
+		endpoint: props.Endpoint,
+		hub:      hub,
+		key:      &props.AccessKey,
+	}, nil
+}
+
+// GenerateClientAccessUrlOptions represents the options for generating a client access url
+type GenerateClientAccessUrlOptions struct {
+	// UserID is the user ID for the client.
+	UserID string
+
+	// Roles are the roles that the connection with the generated token will have.
+	// Roles give the client initial permissions to leave, join, or publish to groups when using PubSub subprotocol.
+	// Possible role values:
+	// - webpubsub.joinLeaveGroup: the client can join or leave any group.
+	// - webpubsub.sendToGroup: the client can send messages to any group.
+	// - webpubsub.joinLeaveGroup.<group>: the client can join or leave group <group>.
+	// - webpubsub.sendToGroup.<group>: the client can send messages to group <group>.
+	// More info: https://azure.github.io/azure-webpubsub/references/pubsub-websocket-subprotocol#permissions
+	Roles []string
+
+	// ExpirationTimeInMinutes is the number of minutes until the token expires.
+	ExpirationTimeInMinutes *int32
```

Comment:
Is 0 a valid value? If not I guess we can use `int32` and use `0` to indicate unprovided value.

Reply:
Use `0` to indicate default value 60 sounds strange though, let me keep it nullable to see how SDK team suggests

Reply:
`0` is not allowed in service side. Use `0` to mean default value then.